workflow/sync/throttler.go

```diff
@@ -38,12 +40,13 @@ var NamespaceBucket BucketFunc = func(key Key) BucketKey {
 }
 
 type throttler struct {
-	queue       QueueFunc
-	bucketFunc  BucketFunc
-	inProgress  buckets
-	pending     map[BucketKey]*priorityQueue
-	lock        *sync.Mutex
-	parallelism int
+	queue             QueueFunc
+	bucketFunc        BucketFunc
+	inProgress        buckets
+	inProgressNoLimit buckets
```

Comment:
Will this give the user the wrong count?
If Parallelism is set to 10, 5 WFs await the lock. Given time, the user may see 15 WFs are running state.
if N WFs are waiting for the Lock then N+10 workflows are in running state.
Is it expected behavior?

Reply:
Just a proposal: Need to discuss 
Can we completely exclude workflows which have locks on this from parallisam queue and let mutex will control it?
I believe that parallelism and mutex locks tackle the same rate limiting problem.
@tczhao @agilgur5 @terrytangyuan @juliev0

Reply:
> I believe that parallelism and mutex locks tackle the same rate limiting problem.

They do, but there are use-cases where both overlap and you'd want to use both simultaneously.
For instance, I have only the resources to run 10 Workflows, but this set of Workflows cannot run in parallel and so needs a mutex. I still don't want more than 10 in that case

Reply:
> Will this give the user the wrong count?
> If Parallelism is set to 10, 5 WFs await the lock. Given time, the user may see 15 WFs are running state.
> if N WFs are waiting for the Lock then N+10 workflows are in running state.

Oh. I think you're right. With a Workflow-level mutex, it might still be `Pending`, but a template-level mutex would remain in `Running` even if it's waiting on a lock.
So it creates the inverse problem of the issue, although only from a visibility and queue perspective, not from an "actually running" perspective

Reply:
Haven't thought about it in-depth yet, but this might require a new state/phase to correctly implement this, i.e. "Locked" or "Waiting" (or something else better named). Which is not the easiest to implement as it requires changes in _a lot_ of places and almost certainly crosses the threshold into a feature as such

Reply:
>Haven't thought about it in-depth yet, but this might require a new state/phase to correctly implement this, i.e. "Locked" or "Waiting" (or something else better named). Which is not the easiest to implement as it requires changes in a lot of places (see also #11849) and almost certainly crosses the threshold into a feature as such

Change on state machine will be a big change. I won't recommend it. It may affect the stability of the product.
Suggestions:
I would recommend adding a sub-state which is just info to the user.
```yaml
Phase: Running
SubPhase: lockWaiting
```

Reply:
> I won't recommend it. It may affect the stability of the product.

We already have needs for new phases, so it's not new in that sense. But yes, it requires a lot of care to implement correctly (as I myself said in various places).

> I would recommend adding a sub-state which is just info to the user.

Mmmm I wouldn't recommend this:
1. We already have _pseudo_-sub-states in the `message` and in a few other forms (e.g. shutdown, hooked, retried, maybe more). Info to the user can already be provided via the `message` and is already used for that purpose, such as for shutdown
1. "Info to the user" is not enough for this if we're going to change counts
  1. So that effectively becomes a nested state, which is generally a _more_ complex structure than flattened, as well as requires a new spec (as opposed to a new enum for the existing state)
    1. In general, various bugs are due to existing psuedo-sub-states. We should prefer correctness to the state machine rather than workarounds on workarounds; we already have an overreliance on such things like derivations too.
    1. Though if nesting makes more sense correctness-wise or complexity wise, then I'm not against that per se; just against workarounds over correctness